src/databases.py

```diff
@@ -0,0 +1,108 @@
+        }
+
+    def _setup_database_requirer(self, relation_name: str, database_name: str) -> DatabaseRequires:
+        """Set up a DatabaseRequires instance.
+
+        The DatabaseRequires instance is an interface between the charm and various data providers.
+        It handles those relations and emit events to help us abstract these integrations.
+
+        Args:
+            relation_name: Name of the data relation
+            database_name: Name of the database (can be overwritten by the provider)
+
+        Returns:
+            DatabaseRequires object produced by the data_platform_libs.v0.data_interfaces library
+        """
+        database_requirer = DatabaseRequires(
+            self._charm,
+            relation_name=relation_name,
+            database_name=database_name,
+        )
+        self._charm.framework.observe(
+            database_requirer.on.database_created, self._charm.on_config_changed
+        )
+        self._charm.framework.observe(
+            self._charm.on[relation_name].relation_broken, self._charm.on_config_changed
+        )
+        return database_requirer
+
+    def get_uris(self) -> typing.Dict[str, str]:
+        """Compute DatabaseURI and return it.
+
+        Returns:
+            DatabaseURI containing details about the data provider integration
+
+        Raises:
+            InvalidDatabaseRelationDataError: if the database relation has invalid data
+        """
+        db_uris: typing.Dict[str, str] = {}
+
+        # the database requires could not be defined
+        # if get_uris() is called before their initialization
+        if not hasattr(self, "_databases") or not self._databases:
+            return db_uris
+
+        for database, db_requires in self._databases.items():
+            relation_data = list(db_requires.fetch_relation_data().values())
+
+            if not relation_data:
+                continue
+
+            # There can be only one database integrated at a time
+            # see: metadata.yaml
+            data = relation_data[0]
+
+            # Check that the relation data is well formed according to the following json_schema:
+            # https://github.com/canonical/charm-relation-interfaces/blob/main/interfaces/mysql_client/v0/schemas/provider.json
+            if not all(data.get(key) for key in ("endpoints", "username", "password")):
+                logger.warning("Incorrect relation data from the data provider: %s", data)
+                raise InvalidDatabaseRelationDataError(
+                    f"Incorrect relation data from the data provider: {data}"
+                )
+
+            database_name = data.get("database", db_requires.database)
+            endpoint = data["endpoints"].split(",")[0]
+            db_uris[f"{database.upper()}_DB_CONNECT_STRING"] = (
+                f"{database}://"
+                f"{data['username']}:{data['password']}"
+                f"@{endpoint}/{database_name}"
+            )
```

Comment:
Could be interesting if there's a library that modeled the DB connection string and use it instead of simply concatenating

Reply:
I could use [urllib.parse.urlunparse](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlunparse)

Reply:
Actually it's not really simple to use `urlunparse` and that will still need some string concatenating.  
There's the [`URL` class of sqlalchemy](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/engine/url.py#L45) but I don't want to import it just for that.